scripts/qa_crowbarsetup.sh

```diff
@@ -1609,20 +1609,61 @@ function set_proposalvars()
     # Determine if we went through an upgrade
     [ -f /etc/cloudsource ] && export cloudsource=$(</etc/cloudsource)
 
-    wantswift=1
-    [ -z "$want_swift" ] && wantceph=1
-    [[ -n "$wanthyperv" ]] && wantswift= && wantceph= && networkingmode=vlan
+    ### dynamic defaults
+    case "$nodenumber" in
+        0|1)
+            deployswift=
+            deployceph=
+        ;;
+        2)
+            deployswift=1
+            deployceph=
+        ;;
+        *)
+            deployswift=
+            deployceph=1
+        ;;
+    esac
+
+    ### filter (temporarily changing defaults)
+    # F1: hyperV only without swift and ceph
+    [ -n "$wanthyperv" ] && deployswift= && deployceph= && networkingmode=vlan
+
+    ### user requests (can override defaults and filters)
+    case "$want_ceph" in
+        '') ;;
+        0)  deployceph= ;;
+        *)  deployceph=1 ;;
+    esac
+    case "$want_swift" in
+        '') ;;
+        0)  deployswift= ;;
+        *)  deployswift=1 ;;
+    esac
+
+    ### constraints
+    # C1: need at least 3 nodes for ceph
+    if [ "$nodenumber" -lt 3 -a "$deployceph" = 1 ] ; then
+        complain 87 "Ceph needs at least 3 nodes to be deployed. You have ${nodenumber} nodes."
+    fi
+
+    # C2: ceph or swift is only possible with at least one volume
+    [ "$cephvolumenumber" -lt 1 ] && deployswift= && deployceph=
+
+    ### FINAL swift and ceph check
+    if [ -n "$deployswift" -a -n "$deployceph" ] ; then
+        complain 89 "Can not deploy ceph and swift at the same time."
+    fi
+    ### do NOT set/change deployceph or deployswift below this line!
+
+    # Tempest
     wanttempest=
     iscloudver 4plus && wanttempest=1
-    [[ "$want_tempest" = 0 ]] && wanttempest=
-
-    [[ "$nodenumber" -lt 3 || "$cephvolumenumber" -lt 1 ]] && wantceph=
-    # we can not use both swift and ceph as each grabs all disks on a node
-    [[ -n "$wantceph" ]] && wantswift=
-    [[ "$cephvolumenumber" -lt 1 ]] && wantswift=
+    [ "$want_tempest" = 0 ] && wanttempest=
 
+    # Cinder
     if [[ -z "$cinder_conf_volume_type" ]]; then
-        if [[ -n "$wantceph" ]]; then
+        if [[ -n "$deployceph" ]]; then
```

Comment:
The `-n` and quotes are redundant here.  You should either use the POSIX-compatible:

    if [ -n "$foo" ]; then

or the `bash`-specific:

    if [[ $foo ]]; then

and preferably maintain consistent usage of one or the other across the whole codebase.